src/followup_questions/followup_questions.py

```diff
@@ -0,0 +1,92 @@
+import typing
+from typing import Protocol
+
+import tiktoken
+from langchain_core.messages import BaseMessage
+from langchain_core.messages.utils import get_buffer_string
+from langchain_core.prompts import PromptTemplate
+
+from agents.common.utils import filter_messages
+from followup_questions.constants import (
+    DEFAULT_HISTORY_MESSAGES_LIMIT,
+    DEFAULT_TOKEN_LIMIT,
+)
+from followup_questions.prompts import FOLLOW_UP_QUESTIONS_PROMPT
+from initial_questions.inital_questions import IEncoding
+from initial_questions.output_parser import QuestionOutputParser
+from utils.logging import get_logger
+from utils.models import IModel
+
+logger = get_logger(__name__)
+
+
+class IFollowUpQuestionsHandler(Protocol):
+    """Protocol for IFollowUpQuestionsHandler."""
+
+    def generate_questions(self, messages: list[BaseMessage]) -> list[str]:
+        """Generates initial questions given a context with cluster data."""
+        ...
+
+
+class FollowUpQuestionsHandler:
+    """Handler that generates follow-up questions."""
+
+    _chain: typing.Any
+    _model: IModel
+    _template: str
+    _tokenizer: IEncoding
+    _token_limit: int = DEFAULT_TOKEN_LIMIT
+    _message_limit: int = DEFAULT_HISTORY_MESSAGES_LIMIT
+
+    def __init__(
+        self,
+        model: IModel,
+        template: str | None = None,
+        tokenizer: IEncoding | None = None,
+    ) -> None:
+        self._model = model
+        self._template = template or FOLLOW_UP_QUESTIONS_PROMPT
+        prompt = PromptTemplate(
+            template=self._template,
+            input_variables=["history"],
+        )
+        output_parser = QuestionOutputParser()
+        self._chain = prompt | model.llm | output_parser
+        self._tokenizer = tokenizer or tiktoken.encoding_for_model(self._model.name)
+
+    def generate_questions(self, messages: list[BaseMessage]) -> list[str]:
+        """Generates follow-up questions given the conversation history."""
+        if len(messages) == 0:
+            return []
+
+        # filter down the conversation history to limit the token count.
+        history = self._get_filtered_history(messages)
+        # invoke the chain to generate follow-up questions.
+        return self._chain.invoke({"history": history})  # type: ignore
```

Comment:
Is there nothing that can be done to fix this instead of ignoring the type checking error?

Reply:
AFAIK, it cannot be fixed atm.